File: BuilderDesignPattern/CarManufacture_Example/Products/Car.java
package BuilderDesignPattern.CarManufacture_Example.Products;

import BuilderDesignPattern.CarManufacture_Example.Components.Engine;
import BuilderDesignPattern.CarManufacture_Example.Components.GPSNavigator;
import BuilderDesignPattern.CarManufacture_Example.Components.Transmission;
import BuilderDesignPattern.CarManufacture_Example.Components.TripComputer;
import BuilderDesignPattern.CarManufacture_Example.ProductTypes.CarsType;

public class Car extends Product {
	private final String model;
	private final CarsType carType;
	private final int seats;
	private final Engine engine;
	private final Transmission transmission;
	private final TripComputer tripComputer;
	private final GPSNavigator gpsNavigator;
	private double fuel = 0;


	/**
	 * @param carModel
	 * string name of the car.
	 * @param carType
	 * enum for the type of car.
	 * @param seats
	 * int number of seats of car.
	 * @param engine
	 * engine for the engine of car.
	 * @param transmission
	 * transmission for the transmission of car.
	 * @param tripComputer
	 * tripComputer for the tripComputer of car.
	 * @param gpsNavigator
	 * gpsNavigator for the gpsNavigator of car.
	 */
	public Car(String carModel, CarsType carType, int seats, Engine engine, Transmission transmission, TripComputer tripComputer, GPSNavigator gpsNavigator) {
		this.model = carModel;
		this.carType = carType;
		this.seats = seats;
		this.engine = engine;
		this.gpsNavigator = gpsNavigator;
		this.transmission = transmission;
		this.tripComputer = tripComputer;
		if (this.tripComputer != null) {
			this.tripComputer.setCar(this);
		}
	}

	public String getCarModel() {
		return this.model;
	}

	public CarsType getCarType() {
		return this.carType;
	}

	public int getSeats() {
		return this.seats;
	}

	public Engine getEngine() {
		return this.engine;
	}

	public TripComputer getTripComputer() {
		return this.tripComputer;
	}

	public GPSNavigator getGpsNavigator() {
		return this.gpsNavigator;
	}

	public Transmission getTransmission() {
		return this.transmission;
	}

	public double getFuel() {
		return this.fuel;
	}

	public void setFuel(double fuel) {
		this.fuel = fuel;
	}

	public String print() {
		String carString = "\nCar Information \n";
		carString += "Name of car: " + model + "\n";
		carString += "Type of car: " + carType + "\n";
		carString += "Count of seats: " + seats + "\n";
		carString += "Engine: volume - " + engine.getVolume() + "; mileage - " + engine.getMileage() + "\n";
		carString += "Transmission: " + transmission + "\n";
		if (this.tripComputer != null) {
			carString += "Trip Computer: Functional" + "\n";
		} else {
			carString += "Trip Computer: N/A" + "\n";
		}
		if (this.gpsNavigator != null) {
			carString += "GPS Navigator: Functional" + "\n";
		} else {
			carString += "GPS Navigator: N/A" + "\n";
		}
		return carString;
	}
}

File: BuilderDesignPattern/CarManufacture_Example_Python/Products/Car.py
from typing import Final

from BuilderDesignPattern.CarManufacture_Example_Python.Components.Engine import Engine
from BuilderDesignPattern.CarManufacture_Example_Python.Components.GPSNavigator import GPSNavigator
from BuilderDesignPattern.CarManufacture_Example_Python.Components.Transmission import Transmission
from BuilderDesignPattern.CarManufacture_Example_Python.Components.TripComputer import TripComputer
from BuilderDesignPattern.CarManufacture_Example_Python.Products.Product import Product
from BuilderDesignPattern.CarManufacture_Example_Python.ProductsTypes.CarsType import CarsType


class Car(Product):

	def __init__(self, model: str, carsType: CarsType, seat: int, engine: Engine, trans: Transmission, tripComp: TripComputer, gpsNav: GPSNavigator):
		super().__init__()
		self._model: Final[ str ] = model
		self._type: Final[ CarsType ] = carsType
		self._seat: Final[ int ] = seat
		self._engine: Final[ Engine ] = engine
		self._transmission: Final[ Transmission ] = trans
		self._tripComputer: Final[ TripComputer ] = tripComp
		self._gpsNavigator: Final[ GPSNavigator ] = gpsNav
		self._fuel: int = 0
		if self._tripComputer is not None:
			self._tripComputer.setCar(self)

	def getModel(self) -> str:
		return self._model

	def getCarType(self) -> CarsType:
		return self._type

	def getSeat(self) -> int:
		return self._seat

	def getEngine(self) -> Engine:
		return self._engine

	def getTransmission(self) -> Transmission:
		return self._transmission

	def getTripComputer(self) -> TripComputer:
		return self._tripComputer

	def getGPSNavigator(self) -> GPSNavigator:
		return self._gpsNavigator

	def getFuel(self) -> int:
		return self._fuel

	def setFuel(self, fuel: int) -> None:
		self._fuel = fuel

	def __str__(self) -> str:
		trip_comp = "Functional" if self._tripComputer is not None else "N/A"
		gps_nav = "Functional" if self._gpsNavigator is not None else "N/A"

		return (
			f"\nCar Information\n"
			f"Name of car: {self._model}\n"
			f"Type of car: {self._type}\n"
			f"Count of seats: {self._seat}\n"
			f"Engine: volume - {self._engine.getVolume()}; mileage - {self._engine.getMileage()}\n"
			f"Transmission: {self._transmission}\n"
			f"Trip Computer: {trip_comp}\n"
			f"GPS Navigator: {gps_nav}\n"
		)
